File: ggplot_and_plotly/histogram_interactive.R
# This script creates a histogram using (a subset of) 'census_data.csv' using the 'ggplot2' and 'plotly' libraries

# A description of 'census_data.csv' can be found in 'census_data_scrape_and_process.R'.
# To avoid an overwhelming plot, I will be using a subset of 'census_data.csv' consisting of
# all observations for 2010.

# Load packages.
library(data.table)
library(ggplot2)
library(grid)
library(plotly)

# Load data.
# Note that the data is located in the 'visualizations' folder, NOT the 'ggplot' folder.
census_data = data.table(read.csv('census_data.csv', 
                                  header = T, 
                                  stringsAsFactors = F, 
                                  check.names = F))

#####
# SUBSET/PROCESS DATA.
  
  # Create data subset consisting of all 2010 observations.
  census_subset = census_data[, .(Name, `2010`)]
  
  # Remove row containing total population of United States.
  census_subset = census_subset[Name != 'United States']
  
  # Divide population values by 1000000 for easier graph viewing.
  census_subset$`2010` = census_subset$`2010` / 1000000
  
#####
# Plot data.
ggplotly(ggplot(data = census_subset, 
       aes(x = `2010`)) +
  geom_histogram(binwidth = 1.5, 
                 fill = I('darkgreen'), 
                 col = I('black')) +
  theme_bw() +
  scale_y_continuous(breaks = seq(0, 14, by = 1)) + 
  scale_x_discrete(breaks = seq(0, 40, by = 5)) +
  theme(axis.text.x = element_text(color = 'black', 
                                   size = 17), 
        axis.text.y = element_text(color = 'black', 
                                   size = 17),
        axis.title.x = element_text(face = 'bold', 
                                    size = 18, 
                                    vjust = -0.5),
        axis.title.y = element_text(face = 'bold',
                                    size = 18),
        plot.title = element_text(face = 'bold', 
                                  size = 19, 
                                  vjust = 2), 
        panel.grid.minor = element_blank()) +
  ggtitle('Histogram of 2010 US State Populations (in millions)') +
  xlab('2010 Populations (in millions)') +
  ylab('Number of States'))

File: plotly/histogram_interactive.py
# This script creates an interactive histogram using 'census_data.csv'.

# A description of 'census_data.csv' can be found in 'census_data_scrape_and_process.R'.

################
# IMPORT MODULES
################

# Import modules.
import os
import pandas as pd
import numpy as np
import plotly
from plotly.graph_objs import Histogram, Scatter, Layout

#############
# IMPORT DATA
#############

# Set working directory.
# This obviously needs to be changed depending on the computer being used.
os.chdir('D:\\Users\\JBansal\\Documents\\GitHub\\visualizations')

# Load data.
census_data = pd.read_csv('census_data.csv',
                          encoding = 'latin-1')

################
# TRANSFORM DATA
################

# Create data subset consisting of all 2010 observations.
census_subset = census_data[['Name', '2010']]

# Remove row containing total population of United States.
census_subset = census_subset[census_subset['Name'] != 'United States']

# Divide population values by 1000000 for easier graph viewing.
census_subset['2010'] = census_subset['2010'] / 1000000

###########
# PLOT DATA
###########

# Add histogram data.
histogram = Histogram(x = census_subset['2010'],
                      autobinx = False,
                      xbins = dict(start = 0,
                                   end = 40,
                                   size = 3))

# Add layout (with annotations).
# The annotations are done more manually than I'd like...
layout = Layout(title = 'Frequency of 2010 US State Populations (in millions)',
                xaxis = dict(title = '2010 Populations (in millions)'),
                yaxis = dict(title = 'Frequency'),
                font = dict(size = 20),
                annotations = [dict(x = 24.5,
                                    y = 2.5,
                                    text = 'Texas',
                                    showarrow = False),
                               dict(x = 25,
                                    y = 1,
                                    text = '25.15M',
                                    showarrow = True),  
                               dict(x = 37,
                                    y = 2.5,
                                    text = 'California',
                                    showarrow = False),
                               dict(x = 37,
                                    y = 1,
                                    text = '37.25M',
                                    showarrow = True)])

# Plot data.
histogram_data = [histogram,]
plotly.offline.plot(dict(data = histogram_data,
                         layout = layout))








